### PR TITLE
shard: support working in degraded mode

### PR DESCRIPTION
**Pull request by fyrchik**
Refs #1143 .

Let's also discuss whether we need a support for `search`.
Also we may introduce a separate `degraded` mode to prevent slowdown when shard was moved to readonly because of maintenance.

### CODE REVIEW

File: pkg/local_object_storage/blobovnicza/get.go [hunk 1]
```diff
@@ -1,6 +1,7 @@
 package blobovnicza
 
 import (
+	"github.com/nspcc-dev/neo-go/pkg/util/slice"
```
Comment (carpawell) [inline pkg/local_object_storage/blobovnicza/get.go:4]: is it ok to use such util from `neo-go` in the engine of the node?

maybe implement somewhere in that repo?

File: pkg/local_object_storage/blobstor/exists.go [hunk 2]
```diff
@@ -29,20 +33,35 @@ func (r ExistsRes) Exists() bool {
...
-		// TODO: #1143 do smth if err != nil
+		var smallErr error
 
-		// check presence in blobovnicza
-		exists, err = b.existsSmall(prm.addr)
+		exists, smallErr = b.existsSmall(prm.addr)
+		if err != nil && (smallErr != nil || exists) {
+			b.log.Warn("error occured during object existence checking",
+				zap.Stringer("address", prm.addr),
+				zap.String("error", err.Error()))
```
Comment (carpawell) [inline pkg/local_object_storage/blobstor/exists.go:53]: `zap.Error(err)`?

File: pkg/local_object_storage/blobstor/exists.go [hunk 3]
```diff
@@ -55,8 +74,36 @@ func (b *BlobStor) existsBig(addr *addressSDK.Address) (bool, error) {
...
+		dirPath := filepath.Dir(p)
+
+		_, ok := activeCache[dirPath]
+
+		_, err := b.getObjectFromLevel(prm, p, !ok)
+		if err != nil {
+			if !errors.Is(err, object.ErrNotFound) {
+				b.log.Debug("could not get object from level",
+					zap.String("level", p),
+					zap.String("error", err.Error()))
```
Comment (carpawell) [inline pkg/local_object_storage/blobstor/exists.go:99]: `zap.Error(err)`?

File: pkg/local_object_storage/blobstor/exists.go [hunk 4]
```diff
@@ -55,8 +74,36 @@ func (b *BlobStor) existsBig(addr *addressSDK.Address) (bool, error) {
...
+	var found bool
+	err := b.iterateSortedLeaves(addr, func(p string) (bool, error) {
+		dirPath := filepath.Dir(p)
+
+		_, ok := activeCache[dirPath]
+
+		_, err := b.getObjectFromLevel(prm, p, !ok)
+		if err != nil {
+			if !errors.Is(err, object.ErrNotFound) {
+				b.log.Debug("could not get object from level",
```
Comment (carpawell) [inline pkg/local_object_storage/blobstor/exists.go:97]: why `debug`?

File: pkg/local_object_storage/shard/exists.go [hunk 5]
```diff
@@ -34,13 +36,26 @@ func (p *ExistsRes) Exists() bool {
 // Returns any error encountered that does not allow to
 // unambiguously determine the presence of an object.
 func (s *Shard) Exists(prm *ExistsPrm) (*ExistsRes, error) {
-	exists, err := s.objectExists(prm.addr)
+	exists, err := meta.Exists(s.metaBase, prm.addr)
+	if err != nil {
+		// If the shard is in degraded mode, try to consult blobstor directly.
+		// Otherwise, just return an error.
+		if s.GetMode() == ModeReadOnly {
```
Comment (carpawell) [inline pkg/local_object_storage/shard/exists.go:43]: why do you try reading only in readOnly mode?

**Review by carpawell — COMMENTED** [review]
> Also we may introduce a separate degraded mode to prevent slowdown

IMO it is important and should be implemented.

File: pkg/local_object_storage/blobovnicza/get.go (hunk 1 above)
Comment (fyrchik) [reply pkg/local_object_storage/blobovnicza/get.go:4]: Why not? It is not that we add a new dependency like in the `api-go` case.

File: pkg/local_object_storage/shard/exists.go (hunk 5 above)
Comment (fyrchik) [reply pkg/local_object_storage/shard/exists.go:43]: That is the whole point of this PR: enable additional logic in degraded mode.

File: pkg/local_object_storage/blobstor/exists.go (hunk 2 above)
Comment (fyrchik) [reply pkg/local_object_storage/blobstor/exists.go:53]: We mostly use `zap.String` in engine. `zap.Error` is more verbose.

**Review by alexvanin — APPROVED** [review]

File: pkg/local_object_storage/blobstor/exists.go (hunk 4 above)
Comment (fyrchik) [reply pkg/local_object_storage/blobstor/exists.go:96]: It is logged in `Exists` if it needs to be reported.

**Pull request merged**